### ISSUE TITLE
jinja2.exceptions.TemplateNotFound

### ISSUE THREAD

**Issue by nhxyz** [niccokunzmann/open-web-calendar#86]
I installed all the dependencies in the requirements.txt but I get this error after starting the app.py and open the web app. Any idea what step I'm missing?

```
jinja2.exceptions.TemplateNotFound
jinja2.exceptions.TemplateNotFound: calendars\dhtmlx.html

Traceback (most recent call last)
File "C:\Users\x\Documents\CalendarApp\ENV\Lib\site-packages\flask\app.py", line 2309, in __call__
return self.wsgi_app(environ, start_response)
File "C:\Users\x\Documents\CalendarApp\ENV\Lib\site-packages\flask\app.py", line 2295, in wsgi_app
response = self.handle_exception(e)
File "C:\Users\x\Documents\CalendarApp\ENV\Lib\site-packages\flask\app.py", line 1741, in handle_exception
reraise(exc_type, exc_value, tb)
File "C:\Users\x\Documents\CalendarApp\ENV\Lib\site-packages\flask\_compat.py", line 35, in reraise
raise value
File "C:\Users\x\Documents\CalendarApp\ENV\Lib\site-packages\flask\app.py", line 2292, in wsgi_app
response = self.full_dispatch_request()
File "C:\Users\x\Documents\CalendarApp\ENV\Lib\site-packages\flask\app.py", line 1815, in full_dispatch_request
rv = self.handle_user_exception(e)
File "C:\Users\x\Documents\CalendarApp\ENV\Lib\site-packages\flask\app.py", line 1718, in handle_user_exception
reraise(exc_type, exc_value, tb)
File "C:\Users\x\Documents\CalendarApp\ENV\Lib\site-packages\flask\_compat.py", line 35, in reraise
```

---
<small>
We're using [Polar.sh](https://polar.sh/niccokunzmann) so you can upvote and help fund this issue. We receive the funding once the issue is completed & confirmed by you. Thank you in advance for helping prioritize & fund our work.
</small>

<a href="https://polar.sh/niccokunzmann/open-web-calendar/issues/86">
<picture>
  <source media="(prefers-color-scheme: dark)" srcset="https://polar.sh/api/github/niccokunzmann/open-web-calendar/issues/86/pledge.svg?darkmode=1">
  <img alt="Fund with Polar" src="https://polar.sh/api/github/niccokunzmann/open-web-calendar/issues/86/pledge.svg">
</picture>
</a>

**Comment by nhxyz** [comment]
Thanks for looking into it. I'm not sure which part I hope this is it.

```python
# constants
HERE = os.path.dirname(__name__) or "."
DEFAULT_SPECIFICATION_PATH = os.path.join(HERE, "default_specification.yml")
TEMPLATE_FOLDER_NAME = "templates"
TEMPLATE_FOLDER = os.path.join(HERE, TEMPLATE_FOLDER_NAME)
CALENDARS_TEMPLATE_FOLDER_NAME = "calendars"
CALENDAR_TEMPLATE_FOLDER = os.path.join(TEMPLATE_FOLDER, CALENDARS_TEMPLATE_FOLDER_NAME)
STATIC_FOLDER_NAME = "static"
STATIC_FOLDER_PATH = os.path.join(HERE, STATIC_FOLDER_NAME)
DHTMLX_LANGUAGES_FILE = os.path.join(STATIC_FOLDER_PATH, "js", "dhtmlx", "locale", "languages.json")

# specification
PARAM_SPECIFICATION_URL = "specification_url"

# globals
app = Flask(__name__, template_folder="templates")
# Check Configuring Flask-Cache section for more details
cache = Cache(app, config={
    'CACHE_TYPE': 'filesystem',
    'CACHE_DIR': tempfile.mktemp(prefix="cache-")})
```

**Comment by niccokunzmann** [comment]
CALENDAR_TEMPLATE_FOLDER would be it. Are there folders there that are named like that?

Am 25. Februar 2022 09:31:45 GMT+00:00 schrieb nhxyz ***@***.***>:

**Comment by nhxyz** [comment]
Yeah it does seem like the paths in the app.py folder look correct. My folder structure should be identical to the repo it looks like this:

```
root folder/
    app.py
    templates/
        about.html
        calendars/
            dhtmlx.html
```

If it's helpful I'll post the errors in the command line. I pasted the error from the web interface initially.

```cmd
(ENV) c:\Users\x\Documents\CalendarApp>python app.py
 * Serving Flask app "app" (lazy loading)
 * Environment: production
   WARNING: Do not use the development server in a production environment.
   Use a production WSGI server instead.
 * Debug mode: on
 * Restarting with stat
 * Debugger is active!
 * Debugger PIN: 310-650-563
 * Running on http://127.0.0.1:5000/ (Press CTRL+C to quit)
127.0.0.1 - - [28/Feb/2022 11:05:48] "GET / HTTP/1.1" 304 -
127.0.0.1 - - [28/Feb/2022 11:05:49] "GET /configuration.js HTTP/1.1" 200 -
127.0.0.1 - - [28/Feb/2022 11:05:49] "GET /calendar.html?url=https%3A%2F%2Fwww.calendarlabs.com%2Fical-calendar%2Fics%2F46%2FGermany_Holidays.ics&language=en&tab=month HTTP/1.1" 500 -
Traceback (most recent call last):
  File "C:\Users\x\Documents\CalendarApp\ENV\lib\site-packages\flask\app.py", line 2309, in __call__
    return self.wsgi_app(environ, start_response)
  File "C:\Users\x\Documents\CalendarApp\ENV\lib\site-packages\flask\app.py", line 2295, in wsgi_app
    response = self.handle_exception(e)
  File "C:\Users\x\Documents\CalendarApp\ENV\lib\site-packages\flask\app.py", line 1741, in handle_exception
    reraise(exc_type, exc_value, tb)
  File "C:\Users\x\Documents\CalendarApp\ENV\lib\site-packages\flask\_compat.py", line 35, in reraise
    raise value
  File "C:\Users\x\Documents\CalendarApp\ENV\lib\site-packages\flask\app.py", line 2292, in wsgi_app
    response = self.full_dispatch_request()
  File "C:\Users\x\Documents\CalendarApp\ENV\lib\site-packages\flask\app.py", line 1815, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "C:\Users\x\Documents\CalendarApp\ENV\lib\site-packages\flask\app.py", line 1718, in handle_user_exception
    reraise(exc_type, exc_value, tb)
  File "C:\Users\x\Documents\CalendarApp\ENV\lib\site-packages\flask\_compat.py", line 35, in reraise
    raise value
  File "C:\Users\x\Documents\CalendarApp\ENV\lib\site-packages\flask\app.py", line 1813, in full_dispatch_request
    rv = self.dispatch_request()
  File "C:\Users\x\Documents\CalendarApp\ENV\lib\site-packages\flask\app.py", line 1799, in dispatch_request
    return self.view_functions[rule.endpoint](**req.view_args)
  File "c:\Users\x\Documents\CalendarApp\app.py", line 156, in get_calendar
    return render_app_template(template, specification)
  File "c:\Users\x\Documents\CalendarApp\app.py", line 132, in render_app_template
    get_query_string=get_query_string
  File "C:\Users\x\Documents\CalendarApp\ENV\lib\site-packages\flask\templating.py", line 134, in render_template
    return _render(ctx.app.jinja_env.get_or_select_template(template_name_or_list),
  File "C:\Users\x\Documents\CalendarApp\ENV\lib\site-packages\jinja2\environment.py", line 930, in get_or_select_template
    return self.get_template(template_name_or_list, parent, globals)
  File "C:\Users\x\Documents\CalendarApp\ENV\lib\site-packages\jinja2\environment.py", line 883, in get_template
    return self._load_template(name, self.make_globals(globals))
  File "C:\Users\x\Documents\CalendarApp\ENV\lib\site-packages\jinja2\environment.py", line 857, in _load_template
    template = self.loader.load(self, name, globals)
  File "C:\Users\x\Documents\CalendarApp\ENV\lib\site-packages\jinja2\loaders.py", line 115, in load
    source, filename, uptodate = self.get_source(environment, name)
  File "C:\Users\x\Documents\CalendarApp\ENV\lib\site-packages\flask\templating.py", line 58, in get_source
    return self._get_source_fast(environment, template)
  File "C:\Users\x\Documents\CalendarApp\ENV\lib\site-packages\flask\templating.py", line 86, in _get_source_fast
    raise TemplateNotFound(template)
jinja2.exceptions.TemplateNotFound: calendars\dhtmlx.html
127.0.0.1 - - [28/Feb/2022 11:05:49] "GET /calendar.html?__debugger__=yes&cmd=resource&f=style.css HTTP/1.1" 200 -
127.0.0.1 - - [28/Feb/2022 11:05:49] "GET /calendar.html?__debugger__=yes&cmd=resource&f=jquery.js HTTP/1.1" 200 -
127.0.0.1 - - [28/Feb/2022 11:05:49] "GET /calendar.html?__debugger__=yes&cmd=resource&f=debugger.js HTTP/1.1" 200 -
127.0.0.1 - - [28/Feb/2022 11:05:50] "GET /calendar.html?__debugger__=yes&cmd=resource&f=console.png HTTP/1.1" 200 -
127.0.0.1 - - [28/Feb/2022 11:05:50] "GET /calendar.html?__debugger__=yes&cmd=resource&f=ubuntu.ttf HTTP/1.1" 200 -
```

**Comment by niccokunzmann** [comment]
I wonder: Where in the code is the template served? Which paths does it access? As I remember, flask has an argument for the template folder. Can you check if this is given somewhere? How is the template rendered?

Am 28. Februar 2022 10:14:16 GMT+00:00 schrieb nhxyz ***@***.***>:

**Comment by niccokunzmann** [comment]
I just looked through the code and saw this:

app = Flask(__name__, template_folder="templates")

If you replace the string with TEMPLATE_FOLDER

app = Flask(__name__, template_folder=TEMPLATE_FOLDER)

Does it change?

It should be changed either way, if you would like to make a pull request.

Am 28. Februar 2022 10:14:16 GMT+00:00 schrieb nhxyz ***@***.***>:

**Comment by Proliecan** [comment]
I tried it and still got the error.

> I just looked through the code and saw this: app = Flask(__name__, template_folder="templates") If you replace the string with TEMPLATE_FOLDER app = Flask(__name__, template_folder=TEMPLATE_FOLDER) Does it change? It should be changed either way, if you would like to make a pull request. Am 28. Februar 2022 10:14:16 GMT+00:00 schrieb nhxyz ***@***.***>:
> […](#)
> Yeah it does seem like the paths in the app.py folder look correct. My folder structure should be identical to the repo it looks like this: ``` root folder/ app.py templates/ about.html calendars/ dhtmlx.html ``` If it's helpful I'll post the errors in the command line. I pasted the error from the web interface initially. ```cmd (ENV) c:\Users\x\Documents\CalendarApp>python app.py * Serving Flask app "app" (lazy loading) * Environment: production WARNING: Do not use the development server in a production environment. Use a production WSGI server instead. * Debug mode: on * Restarting with stat * Debugger is active! * Debugger PIN: 310-650-563 * Running on http://127.0.0.1:5000/ (Press CTRL+C to quit) 127.0.0.1 - - [28/Feb/2022 11:05:48] "GET / HTTP/1.1" 304 - 127.0.0.1 - - [28/Feb/2022 11:05:49] "GET /configuration.js HTTP/1.1" 200 - 127.0.0.1 - - [28/Feb/2022 11:05:49] "GET /calendar.html?url=https%3A%2F%2Fwww.calendarlabs.com%2Fical-calendar%2Fics%2F46%2FGermany_Holidays.ics&language=en&tab=month HTTP/1.1" 500 - Traceback (most recent call last): File "C:\Users\x\Documents\CalendarApp\ENV\lib\site-packages\flask\app.py", line 2309, in __call__ return self.wsgi_app(environ, start_response) File "C:\Users\x\Documents\CalendarApp\ENV\lib\site-packages\flask\app.py", line 2295, in wsgi_app response = self.handle_exception(e) File "C:\Users\x\Documents\CalendarApp\ENV\lib\site-packages\flask\app.py", line 1741, in handle_exception reraise(exc_type, exc_value, tb) File "C:\Users\x\Documents\CalendarApp\ENV\lib\site-packages\flask\_compat.py", line 35, in reraise raise value File "C:\Users\x\Documents\CalendarApp\ENV\lib\site-packages\flask\app.py", line 2292, in wsgi_app response = self.full_dispatch_request() File "C:\Users\x\Documents\CalendarApp\ENV\lib\site-packages\flask\app.py", line 1815, in full_dispatch_request rv = self.handle_user_exception(e) File "C:\Users\x\Documents\CalendarApp\ENV\lib\site-packages\flask\app.py", line 1718, in handle_user_exception reraise(exc_type, exc_value, tb) File "C:\Users\x\Documents\CalendarApp\ENV\lib\site-packages\flask\_compat.py", line 35, in reraise raise value File "C:\Users\x\Documents\CalendarApp\ENV\lib\site-packages\flask\app.py", line 1813, in full_dispatch_request rv = self.dispatch_request() File "C:\Users\x\Documents\CalendarApp\ENV\lib\site-packages\flask\app.py", line 1799, in dispatch_request return self.view_functions[rule.endpoint](**req.view_args) File "c:\Users\x\Documents\CalendarApp\app.py", line 156, in get_calendar return render_app_template(template, specification) File "c:\Users\x\Documents\CalendarApp\app.py", line 132, in render_app_template get_query_string=get_query_string File "C:\Users\x\Documents\CalendarApp\ENV\lib\site-packages\flask\templating.py", line 134, in render_template return _render(ctx.app.jinja_env.get_or_select_template(template_name_or_list), File "C:\Users\x\Documents\CalendarApp\ENV\lib\site-packages\jinja2\environment.py", line 930, in get_or_select_template return self.get_template(template_name_or_list, parent, globals) File "C:\Users\x\Documents\CalendarApp\ENV\lib\site-packages\jinja2\environment.py", line 883, in get_template return self._load_template(name, self.make_globals(globals)) File "C:\Users\x\Documents\CalendarApp\ENV\lib\site-packages\jinja2\environment.py", line 857, in _load_template template = self.loader.load(self, name, globals) File "C:\Users\x\Documents\CalendarApp\ENV\lib\site-packages\jinja2\loaders.py", line 115, in load source, filename, uptodate = self.get_source(environment, name) File "C:\Users\x\Documents\CalendarApp\ENV\lib\site-packages\flask\templating.py", line 58, in get_source return self._get_source_fast(environment, template) File "C:\Users\x\Documents\CalendarApp\ENV\lib\site-packages\flask\templating.py", line 86, in _get_source_fast raise TemplateNotFound(template) jinja2.exceptions.TemplateNotFound: calendars\dhtmlx.html 127.0.0.1 - - [28/Feb/2022 11:05:49] "GET /calendar.html?__debugger__=yes&cmd=resource&f=style.css HTTP/1.1" 200 - 127.0.0.1 - - [28/Feb/2022 11:05:49] "GET /calendar.html?__debugger__=yes&cmd=resource&f=jquery.js HTTP/1.1" 200 - 127.0.0.1 - - [28/Feb/2022 11:05:49] "GET /calendar.html?__debugger__=yes&cmd=resource&f=debugger.js HTTP/1.1" 200 - 127.0.0.1 - - [28/Feb/2022 11:05:50] "GET /calendar.html?__debugger__=yes&cmd=resource&f=console.png HTTP/1.1" 200 - 127.0.0.1 - - [28/Feb/2022 11:05:50] "GET /calendar.html?__debugger__=yes&cmd=resource&f=ubuntu.ttf HTTP/1.1" 200 - ``` -- Reply to this email directly or view it on GitHub: [#86 (comment)](https://github.com/niccokunzmann/open-web-calendar/issues/86#issuecomment-1054097549) You are receiving this because you commented. Message ID: ***@***.***>

**Comment by niccokunzmann** [comment]
Hm. That is funny. I do not have Windows running so I can not test it.
I have two ideas:
1. Can you print the value of "TEMPLATE_FOLDER"? It should be ""C:\Users\x\Documents\CalendarApp\templates", right?
2. In "C:\Users\x\Documents\CalendarApp\ENV\lib\site-packages\flask\templating.py", line 86
This is the line with the error:
You , in _get_source_fast raise TemplateNotFound(template)
I would modify the code in that file to print the relevant variables and their paths? There is probably an if statement testing the existence of the path or a try except. That should clarify what happens a bit better. Maybe the code has a problem with constructing the correct path.

Am 10. Juni 2022 22:39:02 GMT+01:00 schrieb Proliecan ***@***.***>:

**Comment by niccokunzmann** [comment]
I made a quick fix in #118. Can you check out the latest version? The `.\` should not really be used.

**Comment by gagnfra** [comment]
@niccokunzmann Just tested it and it doesn't work. 

I also replaced line 40 to 
app = Flask(__name__, template_folder=TEMPLATE_FOLDER)

and no chances. I think it's around line 164 where it breaks. 

   template_name = specification["template"]
        all_template_names = os.listdir(CALENDAR_TEMPLATE_FOLDER)
        assert template_name in all_template_names, "Template names must be file names like \"{}\", not \"{}\".".format("\", \"".join(all_template_names), template_name)
        template = os.path.join(CALENDARS_TEMPLATE_FOLDER_NAME, template_name)
        return render_app_template(template, specification)

**Comment by gagnfra** [comment]
Or maybe it's something like this (forward slashes instead of backward) https://github.com/bamos/cv/issues/11#issuecomment-76119557

**Comment by gagnfra** [comment]
Found It ! I replaced on line 30 of lib\jinja\loaders.py by 
    for piece in template.split("\\"):
    
It all works now

**Comment by niccokunzmann** [comment]
Uh interesting. Which version of jinja are you using? They should  know..

**Comment by niccokunzmann** [comment]
I reported it and I think, we can close this issue when https://github.com/pallets/jinja/issues/1765 is closed and the version if jinja is in the requirements.txt file.

**Comment by niccokunzmann** [comment]
OK. The problem is that templates can only contain `/` and never `\`. So, the code needs changing a little.

**Comment by niccokunzmann** [comment]
This seems to provide a fix: https://github.com/hsg/simple-calendar-dasboard/commit/e9bde6325ffb35f356b97d6deeb4fd088126397c

**Comment by niccokunzmann** [comment]
I will close this now because a fix is probably provided in #156. If this is not the case, please report it here or in another issue!
